File: go/breathfirst/breathfirst.go
package main

import "fmt"


type Node struct {
	Value string
	Children []*Node
}


// Time -> 0(v + e) Space -> 0(v)
func (n *Node) BreathFirst(array []string) []string{
	queue := make([]*Node, 0)
	queue = append(queue, n)
	for len(queue) > 0 {
		node := queue[0]
		queue = queue[1:]
		array = append(array, node.Value)
		if len(node.Children) > 0 {
			queue = append(queue, node.Children...)
		}
	}

	return array
}


func main() {
	children := []*Node{
		{Value: "B", Children: []*Node{
			{Value: "E"},
			{Value: "F"},
		}},
		{Value: "C"},
		{Value: "D"},
	}
	node := Node{
		Value: "A",
		Children: children,
	}

	fmt.Println(node.BreathFirst([]string{}))
}

File: python/breathfirst.py
class Node():
    def __init__(self, value, children=[]):
        self.value = value
        self.children = children
        

# Time -> 0(v+e), Space -> 0(v) 
def breathfirst(array, node):
    queue = []
    queue.append(node)
    
    while len(queue) > 0:
        node = queue[0]
        queue = queue[1:]
        array.append(node.value)
        if len(node.children) > 0:
            for val in node.children:
                queue.append(val)
    
    return array


children = [Node(
                value="B",
                children=[Node(value="E"), Node(value="F")]
            ),
            Node(
                value="C"
            ),
            Node(value="D")]

node = Node(
    value="A",
    children=children
)
print(breathfirst([], node))